benchmark/src/main/scala/fs2/benchmark/TextBenchmark.scala

```diff
@@ -49,4 +59,12 @@ class TextBenchmark {
       .last
       .unsafeRunSync()
       .get
+
+  @Benchmark
+  def asciiEncode(): Array[Byte] =
```

Comment:
Results from running `benchmark/jmh:run -i 3 -wi 2 -f1 -t1 fs2.benchmark.TextBenchmark` on a M1 Macbook with Temurin 17.

Current PR code:

```
[info] Benchmark                  (asciiStringSize)  (charsetName)   Mode  Cnt       Score       Error  Units
[info] TextBenchmark.asciiEncode                128          utf-8  thrpt    3  122702,219 ± 13773,670  ops/s
[info] TextBenchmark.asciiEncode                128    iso-2022-kr  thrpt    3  119576,150 ±  2410,238  ops/s
[info] TextBenchmark.asciiEncode               1024          utf-8  thrpt    3   23868,525 ±   233,647  ops/s
[info] TextBenchmark.asciiEncode               1024    iso-2022-kr  thrpt    3   23077,995 ±   150,249  ops/s
[info] TextBenchmark.asciiEncode               4096          utf-8  thrpt    3    6537,079 ±   148,060  ops/s
[info] TextBenchmark.asciiEncode               4096    iso-2022-kr  thrpt    3    6092,286 ±   320,573  ops/s
[info] TextBenchmark.asciiEncode              16384          utf-8  thrpt    3    1648,121 ±    23,967  ops/s
[info] TextBenchmark.asciiEncode              16384    iso-2022-kr  thrpt    3    1542,366 ±    10,599  ops/s
[info] TextBenchmark.asciiEncode             131072          utf-8  thrpt    3     190,055 ±    10,821  ops/s
[info] TextBenchmark.asciiEncode             131072    iso-2022-kr  thrpt    3     186,668 ±    25,596  ops/s
```

Slight memory optimization from [comment](https://github.com/typelevel/fs2/pull/3008/files#r990769246):

```
[info] Benchmark                  (asciiStringSize)  (charsetName)   Mode  Cnt       Score      Error  Units
[info] TextBenchmark.asciiEncode                128          utf-8  thrpt    3  124311,351 ± 4393,854  ops/s
[info] TextBenchmark.asciiEncode                128    iso-2022-kr  thrpt    3  123602,921 ± 8403,831  ops/s
[info] TextBenchmark.asciiEncode               1024          utf-8  thrpt    3   23800,856 ± 5362,019  ops/s
[info] TextBenchmark.asciiEncode               1024    iso-2022-kr  thrpt    3   23402,259 ± 1177,082  ops/s
[info] TextBenchmark.asciiEncode               4096          utf-8  thrpt    3    6412,433 ±   42,169  ops/s
[info] TextBenchmark.asciiEncode               4096    iso-2022-kr  thrpt    3    6258,803 ±  384,576  ops/s
[info] TextBenchmark.asciiEncode              16384          utf-8  thrpt    3    1616,414 ±   50,169  ops/s
[info] TextBenchmark.asciiEncode              16384    iso-2022-kr  thrpt    3    1539,644 ±   53,911  ops/s
[info] TextBenchmark.asciiEncode             131072          utf-8  thrpt    3     183,273 ±    8,099  ops/s
[info] TextBenchmark.asciiEncode             131072    iso-2022-kr  thrpt    3     194,158 ±    9,477  ops/s
```

3.3.0 implementation:

```
[info] Benchmark                  (asciiStringSize)  (charsetName)   Mode  Cnt       Score       Error  Units
[info] TextBenchmark.asciiEncode                128          utf-8  thrpt    3  459940,502 ± 62479,321  ops/s
[info] TextBenchmark.asciiEncode                128    iso-2022-kr  thrpt    3  313636,796 ± 84643,728  ops/s
[info] TextBenchmark.asciiEncode               1024          utf-8  thrpt    3  167493,377 ± 16709,846  ops/s
[info] TextBenchmark.asciiEncode               1024    iso-2022-kr  thrpt    3   74485,196 ±  2612,422  ops/s
[info] TextBenchmark.asciiEncode               4096          utf-8  thrpt    3   52266,272 ±   714,696  ops/s
[info] TextBenchmark.asciiEncode               4096    iso-2022-kr  thrpt    3   21595,360 ±   198,851  ops/s
[info] TextBenchmark.asciiEncode              16384          utf-8  thrpt    3   15428,697 ±   240,052  ops/s
[info] TextBenchmark.asciiEncode              16384    iso-2022-kr  thrpt    3    5746,304 ±   126,641  ops/s
[info] TextBenchmark.asciiEncode             131072          utf-8  thrpt    3    1842,642 ±    24,347  ops/s
[info] TextBenchmark.asciiEncode             131072    iso-2022-kr  thrpt    3     631,334 ±    33,201  ops/s
```

So 3x-10x slower on UTF-8, iso-2022-kr is 3x-4x. Not cool 😞 

`getBytes` for UTF-8 seems impossible to reach as the JVM has a lot of performance hacks, boiling down to `System.arraycopy` for ASCII-compatible strings and alike. But I was hoping for being a lot closer on larger inputs and exotic charsets. Haven't done any profiling (yet), could imagine there's an unfortunate use of NIO buffers that leads to copying hidden somewhere. Would appreciate any hint by people that have more experience!

If the performance stays in that ballpark, then we could differentiate by charset and fast path to the old implementation if it doesn't affect correctness (either by picking charsets known to have no BOM or by picking those known to have one – the Java stdlib doesn't seem to provide this information anywhere).

Reply:
I just pushed a new implementation that implements three strategies chosen by the charset in use, based on a discussion with @armanbilge:
* fastpath UTF-8 and charsets without BOM through the old implementation using `getBytes`
* run some known charsets with BOM through `getBytes` and slice the BOMs away (not yet as fast as I hoped for, may need some optimisation)
* fallback to the original implementation of this PR with `CharsetEncoder` as this is the only method known to be correct

Benchmark results (some have high error bars, my machine might have done something during that time):

PR as it is now with the three strategies:

```
[info] Benchmark                  (asciiStringSize)  (charsetName)   Mode  Cnt       Score       Error  Units
[info] TextBenchmark.asciiEncode                128          utf-8  thrpt    3  241078,756 ±  4214,301  ops/s
[info] TextBenchmark.asciiEncode                128         utf-16  thrpt    3   79392,707 ±  3475,150  ops/s
[info] TextBenchmark.asciiEncode                128    iso-2022-kr  thrpt    3   78863,440 ±  6695,626  ops/s
[info] TextBenchmark.asciiEncode               1024          utf-8  thrpt    3  243723,717 ± 17416,529  ops/s
[info] TextBenchmark.asciiEncode               1024         utf-16  thrpt    3   76299,562 ±  2787,154  ops/s
[info] TextBenchmark.asciiEncode               1024    iso-2022-kr  thrpt    3   68631,477 ± 58520,349  ops/s
[info] TextBenchmark.asciiEncode               4096          utf-8  thrpt    3  220528,726 ±  8569,189  ops/s
[info] TextBenchmark.asciiEncode               4096         utf-16  thrpt    3   69097,673 ±  3413,480  ops/s
[info] TextBenchmark.asciiEncode               4096    iso-2022-kr  thrpt    3   45814,458 ±  2688,061  ops/s
[info] TextBenchmark.asciiEncode              16384          utf-8  thrpt    3  177451,798 ± 34625,941  ops/s
[info] TextBenchmark.asciiEncode              16384         utf-16  thrpt    3   39671,226 ±  6648,031  ops/s
[info] TextBenchmark.asciiEncode              16384    iso-2022-kr  thrpt    3   19581,045 ±  1254,195  ops/s
[info] TextBenchmark.asciiEncode             131072          utf-8  thrpt    3   49748,453 ±  3362,374  ops/s
[info] TextBenchmark.asciiEncode             131072         utf-16  thrpt    3    3248,993 ±   122,217  ops/s
[info] TextBenchmark.asciiEncode             131072    iso-2022-kr  thrpt    3    3443,736 ±    22,870  ops/s
```

3.3.0 implementation:

```
[info] Benchmark                  (asciiStringSize)  (charsetName)   Mode  Cnt       Score        Error  Units
[info] TextBenchmark.asciiEncode                128          utf-8  thrpt    3  383735,294 ±  27890,377  ops/s
[info] TextBenchmark.asciiEncode                128         utf-16  thrpt    3  276145,684 ±   9260,096  ops/s
[info] TextBenchmark.asciiEncode                128    iso-2022-kr  thrpt    3  223451,698 ±  61534,850  ops/s
[info] TextBenchmark.asciiEncode               1024          utf-8  thrpt    3  371536,523 ± 117132,997  ops/s
[info] TextBenchmark.asciiEncode               1024         utf-16  thrpt    3  258765,503 ±  50539,873  ops/s
[info] TextBenchmark.asciiEncode               1024    iso-2022-kr  thrpt    3  183692,517 ±   3334,342  ops/s
[info] TextBenchmark.asciiEncode               4096          utf-8  thrpt    3  347988,866 ±  37393,800  ops/s
[info] TextBenchmark.asciiEncode               4096         utf-16  thrpt    3  165533,415 ±   3172,584  ops/s
[info] TextBenchmark.asciiEncode               4096    iso-2022-kr  thrpt    3  108587,378 ±  17709,535  ops/s
[info] TextBenchmark.asciiEncode              16384          utf-8  thrpt    3  242168,751 ±  39448,685  ops/s
[info] TextBenchmark.asciiEncode              16384         utf-16  thrpt    3   64262,596 ±   3758,674  ops/s
[info] TextBenchmark.asciiEncode              16384    iso-2022-kr  thrpt    3   38383,447 ±   2458,706  ops/s
[info] TextBenchmark.asciiEncode             131072          utf-8  thrpt    3   54766,444 ±   4596,102  ops/s
[info] TextBenchmark.asciiEncode             131072         utf-16  thrpt    3    3754,077 ±    194,735  ops/s
[info] TextBenchmark.asciiEncode             131072    iso-2022-kr  thrpt    3    5880,160 ±     65,146  ops/s
```